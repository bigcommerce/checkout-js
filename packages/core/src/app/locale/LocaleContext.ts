import { CurrencyService, LanguageService } from '@bigcommerce/checkout-sdk';
import { createContext } from 'react';

export interface LocaleContextType {
    language: LanguageService;
    date?: {
        inputFormat: string;
    };
    currency?: CurrencyService;
}

const LocaleContext = createContext<LocaleContextType | undefined>(undefined);

export default LocaleContext;
