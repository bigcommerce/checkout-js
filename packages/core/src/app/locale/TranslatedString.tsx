import React, { Fragment, FunctionComponent } from 'react';

import withLanguage, { WithLanguageProps } from './withLanguage';

export interface TranslatedStringProps {
    id: string;
    data?: any;
}

const TranslatedString: FunctionComponent<TranslatedStringProps & WithLanguageProps> = ({
    data,
    id,
    language,
}) => (
    <Fragment>
        { language.translate(id, data) }
    </Fragment>
);

export default withLanguage(TranslatedString);
