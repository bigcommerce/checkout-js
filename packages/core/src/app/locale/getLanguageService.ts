import { createLanguageService, LanguageConfig, LanguageService } from '@bigcommerce/checkout-sdk';

import { FALLBACK_LOCALE, FALLBACK_TRANSLATIONS } from './translations';

let languageService: LanguageService | undefined;

export default function getLanguageService(): LanguageService {
    languageService =
        languageService ??
        createLanguageService({
            fallbackLocale: FALLBACK_LOCALE,
            fallbackTranslations: FALLBACK_TRANSLATIONS,
        });

    return languageService;
}

export type InitializeLanguageService = typeof initializeLanguageService;

export function initializeLanguageService(config: LanguageConfig): LanguageService {
    languageService = createLanguageService({
        ...config,
        defaultLocale: config.locale,
        defaultTranslations: config.defaultTranslations,
        fallbackLocale: FALLBACK_LOCALE,
        fallbackTranslations: FALLBACK_TRANSLATIONS,
    });

    return languageService;
}
