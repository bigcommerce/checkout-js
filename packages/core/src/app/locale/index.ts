export { default as LocaleContext, LocaleContextType } from './LocaleContext';
export { default as createLocaleContext } from './createLocaleContext';
export { default as getDefaultTranslations } from './getDefaultTranslations';
export { default as getLanguageService, InitializeLanguageService } from './getLanguageService';
export { default as masterpassFormatLocale } from './masterpassFormatLocale';
export { default as withCurrency, WithCurrencyProps } from './withCurrency';
export { default as withLanguage, WithLanguageProps } from './withLanguage';
export { default as withDate, WithDateProps } from './withDate';
export { default as LanguageWindow, isLanguageWindow } from './LanguageWindow';
export { default as LocaleProvider } from './LocaleProvider';
export { default as TranslatedHtml, TranslatedHtmlProps } from './TranslatedHtml';
export { default as TranslatedLink, TranslatedLinkProps } from './TranslatedLink';
export { default as TranslatedString, TranslatedStringProps } from './TranslatedString';
