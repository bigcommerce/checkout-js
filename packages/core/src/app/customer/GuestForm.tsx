import { withFormik, FieldProps, FormikProps } from 'formik';
import React, { memo, useCallback, FunctionComponent, ReactNode } from 'react';
import { object, string } from 'yup';

import { withLanguage, TranslatedHtml, TranslatedString, WithLanguageProps } from '../locale';
import { getPrivacyPolicyValidationSchema, PrivacyPolicyField } from '../privacyPolicy';
import { Button, ButtonVariant } from '../ui/button';
import { BasicFormField, Fieldset, Form, Legend  } from '../ui/form';

import EmailField from './EmailField';
import SubscribeField from './SubscribeField';

export interface GuestFormProps {
    canSubscribe: boolean;
    checkoutButtons?: ReactNode;
    continueAsGuestButtonLabelId: string;
    requiresMarketingConsent: boolean;
    defaultShouldSubscribe: boolean;
    email?: string;
    isLoading: boolean;
    privacyPolicyUrl?: string;
    onChangeEmail(email: string): void;
    onContinueAsGuest(data: GuestFormValues): void;
    onShowLogin(): void;
}

export interface GuestFormValues {
    email: string;
    shouldSubscribe: boolean;
}

const GuestForm: FunctionComponent<GuestFormProps & WithLanguageProps & FormikProps<GuestFormValues>> = ({
    canSubscribe,
    checkoutButtons,
    continueAsGuestButtonLabelId,
    isLoading,
    onChangeEmail,
    onShowLogin,
    privacyPolicyUrl,
    requiresMarketingConsent,
}) => {
    const renderField = useCallback((fieldProps: FieldProps<boolean>) => (
        <SubscribeField
            { ...fieldProps }
            requiresMarketingConsent={ requiresMarketingConsent }
        />
    ), [
        requiresMarketingConsent,
    ]);

    return (
        <Form
            className="checkout-form"
            id="checkout-customer-guest"
            testId="checkout-customer-guest"
        >
            <Fieldset
                legend={
                    <Legend hidden>
                        <TranslatedString id="customer.guest_customer_text" />
                    </Legend>
                }
            >
                <p>
                    <TranslatedHtml id="customer.checkout_as_guest_text" />
                </p>

                <div className="customerEmail-container">
                    <div className="customerEmail-body">
                        <EmailField onChange={ onChangeEmail } />

                        { (canSubscribe || requiresMarketingConsent) && <BasicFormField
                            name="shouldSubscribe"
                            render={ renderField }
                        /> }

                        { privacyPolicyUrl && <PrivacyPolicyField
                            url={ privacyPolicyUrl }
                        /> }
                    </div>

                    <div className="form-actions customerEmail-action">
                        <Button
                            className="customerEmail-button"
                            id="checkout-customer-continue"
                            isLoading={ isLoading }
                            testId="customer-continue-as-guest-button"
                            type="submit"
                            variant={ ButtonVariant.Primary }
                        >
                            <TranslatedString id={ continueAsGuestButtonLabelId } />
                        </Button>
                    </div>
                </div>

                {
                    !isLoading && <p>
                        <TranslatedString id="customer.login_text" />
                        { ' ' }
                        <a
                            data-test="customer-continue-button"
                            id="checkout-customer-login"
                            onClick={ onShowLogin }
                        >
                            <TranslatedString id="customer.login_action" />
                        </a>
                    </p>
                }

                { checkoutButtons }
            </Fieldset>
        </Form>
    );
};

export default withLanguage(withFormik<GuestFormProps & WithLanguageProps, GuestFormValues>({
    mapPropsToValues: ({
        email = '',
        defaultShouldSubscribe = false,
        requiresMarketingConsent,
    }) => ({
        email,
        shouldSubscribe: requiresMarketingConsent ? false : defaultShouldSubscribe,
        privacyPolicy: false,
    }),
    handleSubmit: (values, { props: { onContinueAsGuest } }) => {
        onContinueAsGuest(values);
    },
    validationSchema: ({ language, privacyPolicyUrl }: GuestFormProps & WithLanguageProps) => {
        const email = string()
            .email(language.translate('customer.email_invalid_error'))
            .max(256)
            .required(language.translate('customer.email_required_error'));

        const baseSchema = object({ email });

        if (privacyPolicyUrl) {
            return baseSchema.concat(getPrivacyPolicyValidationSchema({
                isRequired: !!privacyPolicyUrl,
                language,
            }));
        }

        return baseSchema;
    },
})(memo(GuestForm)));
