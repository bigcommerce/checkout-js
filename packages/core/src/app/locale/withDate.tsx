import { createInjectHoc, InjectHoc } from '../common/hoc';

import LocaleContext from './LocaleContext';

export interface WithDateProps {
    date: {
        inputFormat: string;
    };
}

const withDate: InjectHoc<WithDateProps> = createInjectHoc(LocaleContext, {
    displayNamePrefix: 'withDate',
    pickProps: (value, key) => key === 'date' && !!value,
});

export default withDate;
