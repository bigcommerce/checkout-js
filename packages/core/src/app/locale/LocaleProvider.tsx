import { CheckoutService, createCurrencyService, StoreConfig } from '@bigcommerce/checkout-sdk';
import { memoizeOne } from '@bigcommerce/memoize';
import React, { Component, ReactNode } from 'react';

import getLanguageService from './getLanguageService';
import LocaleContext from './LocaleContext';

export interface LocaleProviderProps {
    checkoutService: CheckoutService;
}

export interface LocaleProviderState {
    config?: StoreConfig;
}

class LocaleProvider extends Component<LocaleProviderProps> {
    state: Readonly<LocaleProviderState> = {};

    private languageService = getLanguageService();
    private unsubscribe?: () => void;

    private getContextValue = memoizeOne((config?: StoreConfig) => {
        return {
            currency: config ? createCurrencyService(config) : undefined,
            date: config
                ? {
                      inputFormat: config.inputDateFormat,
                  }
                : undefined,
            language: this.languageService,
        };
    });

    componentDidMount(): void {
        const { checkoutService } = this.props;

        this.unsubscribe = checkoutService.subscribe(
            ({ data }) => {
                this.setState({ config: data.getConfig() });
            },
            ({ data }) => data.getConfig(),
        );
    }

    componentWillUnmount(): void {
        if (this.unsubscribe) {
            this.unsubscribe();
            this.unsubscribe = undefined;
        }
    }

    render(): ReactNode {
        const { children } = this.props;
        const { config } = this.state;

        return (
            <LocaleContext.Provider value={this.getContextValue(config)}>
                {children}
            </LocaleContext.Provider>
        );
    }
}

export default LocaleProvider;
