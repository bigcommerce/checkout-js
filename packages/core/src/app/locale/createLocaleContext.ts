import { createCurrencyService, StoreConfig } from '@bigcommerce/checkout-sdk';

import getLanguageService from './getLanguageService';
import { LocaleContextType } from './LocaleContext';

export default function createLocaleContext(config: StoreConfig): Required<LocaleContextType> {
    if (!config) {
        throw new Error('Missing configuration data');
    }

    const { inputDateFormat } = config;

    return {
        currency: createCurrencyService(config),
        date: {
            inputFormat: inputDateFormat,
        },
        language: getLanguageService(),
    };
}
