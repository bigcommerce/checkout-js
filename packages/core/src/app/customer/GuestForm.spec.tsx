import { mount, render } from 'enzyme';
import React, { FunctionComponent } from 'react';

import { getStoreConfig } from '../config/config.mock';
import { createLocaleContext, LocaleContext, LocaleContextType } from '../locale';
import { PrivacyPolicyField } from '../privacyPolicy';

import GuestForm, { GuestFormProps } from './GuestForm';

describe('GuestForm', () => {
    let defaultProps: GuestFormProps;
    let localeContext: LocaleContextType;
    let TestComponent: FunctionComponent<Partial<GuestFormProps>>;

    beforeEach(() => {
        defaultProps = {
            canSubscribe: true,
            continueAsGuestButtonLabelId: 'customer.continue',
            defaultShouldSubscribe: false,
            isLoading: false,
            onChangeEmail: jest.fn(),
            onContinueAsGuest: jest.fn(),
            onShowLogin: jest.fn(),
            requiresMarketingConsent: false,
        };

        localeContext = createLocaleContext(getStoreConfig());

        TestComponent = (props) => (
            <LocaleContext.Provider value={localeContext}>
                <GuestForm {...defaultProps} {...props} />
            </LocaleContext.Provider>
        );
    });

    it('matches snapshot', () => {
        const component = render(<TestComponent />);

        expect(component).toMatchSnapshot();
    });

    it('renders form with initial values', () => {
        const component = mount(
            <TestComponent defaultShouldSubscribe={true} email="test@bigcommerce.com" />,
        );

        expect(component.find('input[name="email"]').prop('value')).toBe('test@bigcommerce.com');

        expect(component.find('input[name="shouldSubscribe"]').prop('value')).toBe(true);
    });

    it('notifies when user clicks on "continue as guest" button', async () => {
        const handleContinueAsGuest = jest.fn();
        const component = mount(<TestComponent onContinueAsGuest={handleContinueAsGuest} />);

        component
            .find('input[name="email"]')
            .simulate('change', { target: { value: 'test@bigcommerce.com', name: 'email' } });

        component
            .find('input[name="shouldSubscribe"]')
            .simulate('change', { target: { value: true, name: 'shouldSubscribe' } });

        component.find('form').simulate('submit');

        await new Promise((resolve) => process.nextTick(resolve));

        expect(handleContinueAsGuest).toHaveBeenCalledWith({
            email: 'test@bigcommerce.com',
            privacyPolicy: false,
            shouldSubscribe: true,
        });
    });

    it('disables "continue as guest" button when isLoading is true', () => {
        const handleContinueAsGuest = jest.fn();
        const component = mount(
            <TestComponent isLoading={true} onContinueAsGuest={handleContinueAsGuest} />,
        );

        const button = component.find('[data-test="customer-continue-as-guest-button"]');

        expect(button.prop('disabled')).toBeTruthy();
    });

    it('displays error message if email is not valid', () => {
        async function getEmailError(value: string): Promise<string> {
            const handleContinueAsGuest = jest.fn();
            const component = mount(
                <TestComponent email={value} onContinueAsGuest={handleContinueAsGuest} />,
            );

            component.find('form').simulate('submit');

            await new Promise((resolve) => process.nextTick(resolve));

            component.update();

            return component.find('[data-test="email-field-error-message"]').text();
        }

        expect(getEmailError('')).resolves.toBe('Email address is required');

        const invalidEmailMessage = 'Email address must be valid';

        expect(getEmailError('test@bigcommerce')).resolves.toEqual(invalidEmailMessage);
        expect(getEmailError('test')).resolves.toEqual(invalidEmailMessage);
        expect(getEmailError('@test.test')).resolves.toEqual(invalidEmailMessage);
        expect(getEmailError('test@.test')).resolves.toEqual(invalidEmailMessage);
        expect(getEmailError('test@test.')).resolves.toEqual(invalidEmailMessage);
        expect(getEmailError('test@te st.test')).resolves.toEqual(invalidEmailMessage);
        expect(getEmailError('test@test.comtest@test.com')).resolves.toEqual(invalidEmailMessage);
        expect(getEmailError('test@test.com,test@test.com')).resolves.toEqual(invalidEmailMessage);
    });

    it('notifies when user clicks on "sign in" button', () => {
        const handleShowLogin = jest.fn();
        const component = mount(<TestComponent onShowLogin={handleShowLogin} />);

        component.find('[data-test="customer-continue-button"]').simulate('click');

        expect(handleShowLogin).toHaveBeenCalled();
    });

    it('calls "onChangeEmail" handler when user changes email address', () => {
        const handleChangeEmail = jest.fn();
        const component = mount(<TestComponent onChangeEmail={handleChangeEmail} />);

        component
            .find('input[name="email"]')
            .simulate('change', { target: { value: 'test@bigcommerce', name: 'email' } });

        expect(handleChangeEmail).toHaveBeenCalledWith('test@bigcommerce');
    });

    it('calls "onChangeEmail" handler when user changes email address with strict validation enabled and it includes a domain', () => {
        const handleChangeEmail = jest.fn();
        const component = mount(<TestComponent onChangeEmail={handleChangeEmail} />);

        component
            .find('input[name="email"]')
            .simulate('change', { target: { value: 'test@bigcommerce.com', name: 'email' } });

        expect(handleChangeEmail).toHaveBeenCalledWith('test@bigcommerce.com');
    });

    it('renders newsletter field if store allows newsletter subscription', () => {
        const component = mount(<TestComponent canSubscribe={true} />);

        expect(component.find('label[htmlFor="shouldSubscribe"]').text()).toBe(
            'Subscribe to our newsletter.',
        );
        expect(component.exists('input[name="shouldSubscribe"]')).toBe(true);
    });

    it('renders marketing consent field', () => {
        const component = mount(
            <TestComponent canSubscribe={true} requiresMarketingConsent={true} />,
        );

        expect(component.find('label[htmlFor="shouldSubscribe"]').text()).toBe(
            'I would like to receive updates and offers.',
        );
        expect(component.exists('input[name="shouldSubscribe"]')).toBe(true);
    });

    it('sets newsletter field with default value', () => {
        const Container = ({ defaultShouldSubscribe }: { defaultShouldSubscribe: boolean }) => (
            <TestComponent canSubscribe={true} defaultShouldSubscribe={defaultShouldSubscribe} />
        );

        const componentA = mount(<Container defaultShouldSubscribe={true} />);
        const componentB = mount(<Container defaultShouldSubscribe={false} />);

        expect(componentA.find('input[name="shouldSubscribe"]').prop('value')).toBe(true);

        expect(componentB.find('input[name="shouldSubscribe"]').prop('value')).toBe(false);
    });

    it('renders privacy policy field', () => {
        const component = mount(<TestComponent privacyPolicyUrl="foo" />);

        expect(component.find(PrivacyPolicyField)).toHaveLength(1);
    });

    it('displays error message if privacy policy is required and not checked', async () => {
        const handleContinueAsGuest = jest.fn();
        const component = mount(
            <TestComponent onContinueAsGuest={handleContinueAsGuest} privacyPolicyUrl="foo" />,
        );

        component
            .find('input[name="email"]')
            .simulate('change', { target: { value: 'test@test.com', name: 'email' } });

        component.find('form').simulate('submit');

        await new Promise((resolve) => process.nextTick(resolve));

        component.update();

        expect(handleContinueAsGuest).not.toHaveBeenCalled();

        expect(component.find('[data-test="privacy-policy-field-error-message"]').text()).toBe(
            'Please agree to the Privacy Policy.',
        );
    });

    it('does not render "sign in" button when loading', () => {
        const component = mount(<TestComponent isLoading={true} />);

        expect(component.find('[data-test="customer-continue-button"]')).toHaveLength(0);
    });

    it('shows different action button label if another label id was provided', () => {
        const component = mount(<TestComponent continueAsGuestButtonLabelId="customer.continue" />);

        expect(component.find('[data-test="customer-continue-button"]').text()).not.toBe(
            'Continue as guest',
        );
    });
});
