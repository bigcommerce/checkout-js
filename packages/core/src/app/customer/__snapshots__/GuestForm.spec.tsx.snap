// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GuestForm matches snapshot 1`] = `
<div>
  <form
    class="checkout-form"
    data-test="checkout-customer-guest"
    id="checkout-customer-guest"
    novalidate=""
  >
    <fieldset
      class="form-fieldset"
    >
      <legend
        class="form-legend is-srOnly"
      >
        Guest Customer
      </legend>
      <div
        class="form-body"
      >
        <div
          class="customerEmail-container"
        >
          <div
            class="customerEmail-body"
          >
            <div
              class="form-field"
            >
              <label
                class="form-label optimizedCheckout-form-label"
                for="email"
                id="email-label"
              >
                Email
              </label>
              <input
                autocomplete="email"
                class="form-input optimizedCheckout-form-input"
                id="email"
                name="email"
                type="email"
                value=""
              />
            </div>
            <div
              class="form-field"
            >
              <input
                class="form-checkbox"
                id="shouldSubscribe"
                name="shouldSubscribe"
                type="checkbox"
                value="false"
              />
              <label
                class="form-label optimizedCheckout-form-label"
                for="shouldSubscribe"
              >
                Subscribe to our newsletter.
              </label>
            </div>
          </div>
          <div
            class="form-actions customerEmail-action"
          >
            <button
              class="button customerEmail-button button--primary optimizedCheckout-buttonPrimary"
              data-test="customer-continue-as-guest-button"
              id="checkout-customer-continue"
              type="submit"
            >
              Continue
            </button>
          </div>
        </div>
        <p>
          Already have an account? 
          <a
            data-test="customer-continue-button"
            id="checkout-customer-login"
          >
            Sign in now
          </a>
        </p>
      </div>
    </fieldset>
  </form>
</div>
`;
